shared/lib/neighbors/bruteNeighborhood.ts

```diff
@@ -0,0 +1,75 @@
+/**
+*  @license
+* Copyright 2021, JsData. All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { NAryHeap } from './nAryHeap'
+import {
+  Neighbor,
+  Neighborhood,
+  NeighborhoodEntry,
+  NeighborhoodParams
+} from './neighborhood'
+import { Metric } from './metrics'
+
+export class BruteNeighborhood<A, V> implements Neighborhood<A, V> {
+  private _metric: Metric<A>
+  private _entries: NeighborhoodEntry<A, V>[]
+
+  constructor({ metric, entries = [] }: NeighborhoodParams<A, V>) {
+    this._metric = metric
+    this._entries = [
+      ...(function* () {
+        for (const { address, value } of entries) yield { address, value }
+      })()
+    ]
```

Comment:
Is there a reason that we can't just go `this._entries = entries`? Even if you are `yield`ing for the sake of performance, the `entries` object must exist in memory here, so I'm not sure about the speed benefits of this. I could be misunderstanding what the value of this is though.

Reply:
Entries is an Iterable not an Array. It's also intended as a deep protection copy:
  * The caller may do other stuff with the objects and be confused that the behavior of the neighborhood changes
  * In case of a kd-Tree or Ball Tree, the data would be copied into a tree which is consistent with protection copies
  *  The entries might contain many more entries which would become unreachable (memory leaks).